contracts/contracts/Subcall.sol

```diff
@@ -78,11 +92,259 @@ library Subcall {
                 "amount",
                 hex"8250",
                 value,
-                hex"40"
+                uint8(0x40 + token.length),
+                token
             )
         );
     }
 
+    /**
+     * Returns a CBOR encoded structure, containing the following possible keys.
+     * All keys are optional:
+     *
+     *  shares: u128
+     *  epoch: EpochTime
+     *  receipt: u64
+     *  amount: u128
+     *  error: {module: string, code: u32}
+     *
+     * Delegate will have the `error` or `shares` keys.
+     * UndelegateStart will have the `epoch` and `receipt` keys.
+     * UndelegateDone will have the `amount` key
+     *
+     * @param kind 1=Delegate, 2=UndelegateStart, 3=UndelegateDone
+     * @param receiptId ID of receipt
+     */
+    function consensusTakeReceipt(SubcallReceiptKind kind, uint64 receiptId)
+        internal
+        returns (bytes memory)
+    {
+        require(receiptId != 0);
+
+        require(uint8(kind) > 0 && uint8(kind) <= 23);
+
+        (bool success, bytes memory data) = SUBCALL.call(
+            abi.encode(
+                CONSENSUS_TAKE_RECEIPT,
+                abi.encodePacked(
+                    hex"a2", // Map, 2 pairs
+                    hex"62",
+                    "id", // Byte string, 2 bytes
+                    hex"1b",
+                    receiptId, // Unsigned 64bit integer
+                    hex"64",
+                    "kind", // Byte string, 4 bytes
+                    uint8(kind) // uint8 <= 23.
+                )
+            )
+        );
+
+        require(success);
+
+        (uint64 status, bytes memory result) = abi.decode(
+            data,
+            (uint64, bytes)
+        );
+
+        if (status != 0) {
+            revert ConsensusTakeReceiptError(status, string(result));
+        }
+
+        return result;
+    }
+
+    function _parseCBORUint(bytes memory result, uint256 offset)
```

Comment:
Are you sure we want to implement CBOR ourselves? Google returns me this: https://owlprotocol.github.io/solidity-cbor/docs/ but I didn't look at it closely yet.

If yes, then I'd at least move CBOR functions to a separate `utils/Cbor.sol` library or so.

Reply:
It's significantly more gas efficient to just do it this way, versus wrapping everything in an extra layer of CBOR handling library.